pkg/database/all_ops.go

```diff
@@ -120,20 +136,30 @@ func (d *db) ExecAll(req *schema.ExecAllRequest) (*schema.TxHeader, error) {
 					return nil, store.ErrIllegalArguments
 				}
 
-				// zAdd arguments are converted in regular key value items and then atomically inserted
+				if req.NoWait && (x.ZAdd.AtTx != 0 || !x.ZAdd.BoundRef) {
+					return nil, ErrNoWaitOperationMustBeSelfContained
+				}
+
 				_, exists := kmap[sha256.Sum256(x.ZAdd.Key)]
 
-				if !exists || x.ZAdd.AtTx > 0 {
-					// check referenced key exists and it's not a reference
-					refEntry, err := d.getAt(EncodeKey(x.ZAdd.Key), x.ZAdd.AtTx, 0, index, tx)
-					if err != nil {
-						return nil, err
-					}
-					if refEntry.ReferencedBy != nil {
-						return nil, ErrReferencedKeyCannotBeAReference
+				if req.NoWait && !exists {
+					return nil, ErrNoWaitOperationMustBeSelfContained
+				}
```

Comment:
Same as with the `Op_Ref` - we could add some extra checks here to make recursive and cyclic references much harder to add.